### PR TITLE
Introduce Polly.Extensions project

### PR DESCRIPTION
**Pull request by martintmk**
### The issue or feature being addressed

#1087 

### Details on the issue fix or feature implementation

Enables the configuration of `ResilienceStrategy` using the DI and additional extensibility. The API exposes only a single `AddResilienceStrategy` public extension for `IServiceCollecion`. Additional extensibility can be done using `IOptions` pattern.

``` csharp

var services = new ServiceCollection();

// define your strategy
services.AddResilienceStrategy(
  "my-key", 
   context => context.Builder.AddTimeout(TimeSpan.FromSeconds(10)));

// define your strategy using custom options
services.AddResilienceStrategy(
    "my-timeout",
    context =>
    {
        var myOptions = context.ServiceProvider.GetRequiredService<IOptions<MyTimeoutOptions>>().Value;
        context.Builder.AddTimeout(myOptions.Timeout);
    });

// configure global builder options that will be shared for all builders
services.Configure<ResilienceStrategyBuilderOptions>(options =>
{
    options.Properties.Set(new ResiliencePropertyKey<string>("my-key"), "shared-value");
});

// configure the default telemetry factory
services.TryAddSingleton<ResilienceTelemetryFactory, MyTelemetryFactory>();

// use your strategy
var serviceProvider = services.BuildServiceProvider();
var strategyProvider = serviceProvider.GetRequiredService<ResilienceStrategyProvider<string>>();

var resilienceStrategy = strategyProvider.Get("my-key");

// use your strategy
```

Additional Changes:
- using net462 target (net461 is out of support and some packages complained).

### Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have successfully run a local build

### CODE REVIEW

File: src/Polly.Hosting/DependencyInjection/ConfigureResilienceStrategyRegistryOptions.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+namespace Polly.DependencyInjection;
+
+internal class ConfigureResilienceStrategyRegistryOptions<TKey>
```
Comment (martincostello) [inline src/Polly.Hosting/DependencyInjection/ConfigureResilienceStrategyRegistryOptions.cs:3]: Can this be `sealed`, or is there a derived type somewhere?

File: src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+using Microsoft.Extensions.Options;
+using Polly.Builder;
+using Polly.Registry;
+using Polly.Telemetry;
+using Polly.Utils;
```
Comment (martincostello) [inline src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs:9]: Some of these can be tidied due to implicit usings.

File: src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        return services;
+    }
+
+    private static IServiceCollection AddResilienceStrategyRegistry<TKey>(this IServiceCollection services)
+        where TKey : notnull
+    {
+        services.TryAddSingleton(serviceProvider =>
+        {
+            var options = serviceProvider.GetRequiredService<IOptions<ResilienceStrategyRegistryOptions<TKey>>>().Value;
```
Comment (martincostello) [inline src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs:63]: If we depend on options, should we explicitly call `.AddOptions()` (which I think is a no-op if already done?) as part of adding our own registrations? Or do the `AddOptions<T>()` calls already do that?

File: src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            var globalOptions = serviceProvider.GetRequiredService<IOptions<ResilienceStrategyBuilderOptions>>().Value;
+
+            return new ResilienceStrategyBuilder
+            {
+                Options = new ResilienceStrategyBuilderOptions(globalOptions)
+            };
+        });
+    }
+
+    private class RegistryMarker<TKey>
```
Comment (martincostello) [inline src/Polly.Hosting/DependencyInjection/PollyServiceCollectionExtensions.cs:123]:
```suggestion
    private sealed class RegistryMarker<TKey>
```

File: src/Polly.Hosting/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# About Polly.Hosting
+
+The `Polly.Hosting` enables the following features:
+
+
+- Integrates the Polly with the standard `IServiceCollection` DI container.
```
Comment (martincostello) [inline src/Polly.Hosting/README.md:6]:
```suggestion
- Integrates Polly with the standard `IServiceCollection` Dependency Injection (DI) container.
```

File: src/Polly.Hosting/README.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+- Implements the `ResilienceTelemetryFactory` that adds [logging](https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line) and [metering](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/metrics) for all strategies created using the DI extension points.
```
Comment (martincostello) [inline src/Polly.Hosting/README.md:7]:
```suggestion
- Implements `ResilienceTelemetryFactory` that adds [logging](https://learn.microsoft.com/dotnet/core/extensions/logging?tabs=command-line) and [metering](https://learn.microsoft.com/dotnet/core/diagnostics/metrics) for all strategies created using the DI extension points.
```

File: src/Polly.Hosting/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+// use your strategy
+var serviceProvider = services.BuildServiceProvider();
+var strategyProvider = serviceProvider.GetRequiredService<ResilienceStrategyProvider<string>>();
+
+var resilienceStrategy = strategyProvider.Get("my-key");
+
+// use your strategy
+The consumer just calls the AddResilienceStrategy and provides a callback that configures the resilience strategy.
+```
```
Comment (martincostello) [inline src/Polly.Hosting/README.md:37]:
```suggestion
``` csharp
var services = new ServiceCollection();

// Define your strategy
services.AddResilienceStrategy(
  "my-key", 
   context => context.Builder.AddTimeout(TimeSpan.FromSeconds(10)));

// Define your strategy using custom options
services.AddResilienceStrategy(
    "my-timeout",
    context =>
    {
        var myOptions = context.ServiceProvider.GetRequiredService<IOptions<MyTimeoutOptions>>().Value;
        context.Builder.AddTimeout(myOptions.Timeout);
    });

// Use your strategy
var serviceProvider = services.BuildServiceProvider();
var strategyProvider = serviceProvider.GetRequiredService<ResilienceStrategyProvider<string>>();

var resilienceStrategy = strategyProvider.Get("my-key");

// Use your strategy
// The consumer calls AddResilienceStrategy() and provides a callback that configures the resilience strategy.
```
```

File: src/Polly/Polly.csproj [hunk 8]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net472;net461;</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0;net472;net462;</TargetFrameworks>
```
Comment (martincostello) [inline src/Polly/Polly.csproj:4]: What was the motivating factor for the TFM bump?

File: src/Polly.Hosting/Polly.Hosting.csproj [hunk 9]
```diff
@@ -0,0 +1,31 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>net7.0;net6.0;netstandard2.0;net472;net462</TargetFrameworks>
+    <AssemblyTitle>Polly.Hosting</AssemblyTitle>
```
Comment (juraj-blazek) [inline src/Polly.Hosting/Polly.Hosting.csproj:4]: `Polly.Hosting` doesn't sound right => `Polly.Extensions`?
We could then follow the familiar pattern with `Polly.Extensions.DependencyInjection`

**Comment by martincostello** [comment]
Just my two outstanding comments from yesterday.

**Comment by martintmk** [comment]
> Just my two outstanding comments from yesterday.

@martincostello Sorry, but I do not see any that I did not respond to. Are they lost somewhere? :)

**Comment by martincostello** [comment]
This one and the other it turns out wasn't submitted 😅

https://github.com/App-vNext/Polly/pull/1088#discussion_r1150245950

File: src/Polly/Polly.csproj (hunk 8 above)
Comment (martincostello) [reply src/Polly/Polly.csproj:4]: Needed for the [System.Diagnostics.DiagnosticSource](https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/#dependencies-body-tab) package.

**Review by martincostello — APPROVED** [review]

**Pull request merged**